src/protocol/codecs/msgpack.rs

```diff
@@ -382,6 +462,181 @@ impl<'de> Deserialize<'de> for HashDigest {
+pub struct MultisigSignature {
+    #[serde(rename = "subsig")]
+    pub subsigs: Vec<MultisigSubsig>,
+
+    #[serde(rename = "thr")]
+    pub threshold: u8,
+
+    #[serde(rename = "v")]
+    pub version: u8,
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Deserialize)]
+pub struct MultisigSubsig {
+    #[serde(rename = "pk")]
+    pub key: Ed25519PublicKey,
+
+    #[serde(rename = "s")]
+    pub sig: Option<Ed25519Signature>,
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct Ed25519PublicKey(pub [u8; 32]);
+
+impl Serialize for Ed25519PublicKey {
+    fn serialize<S>(&self, serializer: S) -> Result<<S as Serializer>::Ok, <S as Serializer>::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_bytes(&self.0[..])
+    }
+}
+
+impl<'de> Deserialize<'de> for Ed25519PublicKey {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Ok(Ed25519PublicKey(
+            deserializer.deserialize_bytes(VisitorU8_32)?,
+        ))
+    }
+}
+
+/// A Seed holds the entropy needed to generate cryptographic keys.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct Ed25519Seed(pub [u8; 32]);
+
+impl Serialize for Ed25519Seed {
+    fn serialize<S>(&self, serializer: S) -> Result<<S as Serializer>::Ok, <S as Serializer>::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_bytes(&self.0[..])
+    }
+}
+
+impl<'de> Deserialize<'de> for Ed25519Seed {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Ok(Ed25519Seed(deserializer.deserialize_bytes(VisitorU8_32)?))
+    }
+}
+
+/// Verifiable Random Function proof.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct VrfProof(pub [u8; 80]);
+
+impl Serialize for VrfProof {
+    fn serialize<S>(&self, serializer: S) -> Result<<S as Serializer>::Ok, <S as Serializer>::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_bytes(&self.0[..])
+    }
+}
+
+impl<'de> Deserialize<'de> for VrfProof {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Ok(VrfProof(deserializer.deserialize_bytes(VisitorU8_80)?))
+    }
+}
+
+pub struct SignatureVisitor;
+
+impl<'de> Visitor<'de> for SignatureVisitor {
+    type Value = [u8; 64];
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("expecting a 64 byte array")
+    }
+
+    fn visit_bytes<E>(self, v: &[u8]) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        if v.len() != 64 {
+            return Err(E::custom(format!("invalid byte array length: {}", v.len())));
+        }
+
+        let mut bytes = [0; 64];
+        bytes.copy_from_slice(v);
+        Ok(bytes)
+    }
+}
+
+pub struct VisitorU8_80;
+
+impl<'de> Visitor<'de> for VisitorU8_80 {
+    type Value = [u8; 80];
```

Comment:
80 could be some const to avoid magics - here and in other places.

Reply:
Done, see [comment](https://github.com/runziggurat/algorand/pull/42/files#r1045740140)